gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/datafetcher/EntityInteractionsFetcher.java

```diff
@@ -119,31 +130,79 @@ public EntityInteractionsFetcher(
   }
 
   public void populateEntityInteractions(
-      RequestContext context, EntitiesRequest request, Map<EntityKey, Builder> entityBuilders) {
-    // Process the incoming interactions
-    if (!InteractionsRequest.getDefaultInstance().equals(request.getIncomingInteractions())) {
-      addInteractions(
-          context,
-          request,
-          entityBuilders,
-          request.getIncomingInteractions(),
-          INCOMING,
-          "fromEntityType filter is mandatory for incoming interactions.");
+      RequestContext context,
+      EntitiesRequest entitiesRequest,
+      Map<EntityKey, Builder> entityBuilders) {
+    List<EntityInteractionQueryRequest> allQueryRequests = new ArrayList<>();
+    // Process the incoming interactions, and prepare QS queries
+    if (!InteractionsRequest.getDefaultInstance()
+        .equals(entitiesRequest.getIncomingInteractions())) {
+      allQueryRequests.addAll(
+          prepareQueryRequests(
+              context,
+              entitiesRequest,
+              entityBuilders,
+              entitiesRequest.getIncomingInteractions(),
+              INCOMING,
+              "fromEntityType filter is mandatory for incoming interactions."));
     }
 
-    // Process the outgoing interactions
-    if (!InteractionsRequest.getDefaultInstance().equals(request.getOutgoingInteractions())) {
-      addInteractions(
-          context,
-          request,
-          entityBuilders,
-          request.getOutgoingInteractions(),
-          OUTGOING,
-          "toEntityType filter is mandatory for outgoing interactions.");
+    // Process the outgoing interactions, and prepare the QS queries
+    if (!InteractionsRequest.getDefaultInstance()
+        .equals(entitiesRequest.getOutgoingInteractions())) {
+      allQueryRequests.addAll(
+          prepareQueryRequests(
+              context,
+              entitiesRequest,
+              entityBuilders,
+              entitiesRequest.getOutgoingInteractions(),
+              OUTGOING,
+              "toEntityType filter is mandatory for outgoing interactions."));
     }
+
+    // execute all the requests in parallel, and wait for results
+    List<CompletableFuture<EntityInteractionQueryResponse>> queryRequestCompletableFutures =
+        allQueryRequests.stream()
+            .map(
+                e ->
+                    CompletableFuture.supplyAsync(
+                        () -> executeQueryRequest(context, e), this.executorService))
+            .collect(Collectors.toList());
+
+    // wait and parse result as an when complete
+    queryRequestCompletableFutures.forEach(
+        queryRequestCompletableFuture -> {
+          EntityInteractionQueryResponse qsResponse = queryRequestCompletableFuture.join();
+          InteractionsRequest interactionsRequest =
```

Comment:
nit: we already determined this on the request side, would be nice to bring that request through rather than recalcing it.

Reply:
I found it more readable, so taken care of as part of this PR.